kubejs/server_scripts/enigmatica/kubejs/base/recipetypes/minecraft/smithing.js

```diff
@@ -3,10 +3,31 @@ onEvent('recipes', (event) => {
         {
             input1: Item.of('aiotbotania:terra_aiot', '{Damage:0}').weakNBT(),
             input2: 'kubejs:quintuple_alfsteel_ingot',
-            output: 'aiotbotania:alfsteel_aiot'
+            output: 'aiotbotania:alfsteel_aiot',
+            id: `aiotbotania:alfsteel_aiot_smithing`
         }
     ];
+
+    const black_hole_types = ['tank', 'unit'];
+    const black_hole_tiers = ['common', 'pity', 'simple', 'advanced', 'supreme'];
+
+    black_hole_types.forEach(type => {
+        black_hole_tiers.forEach((tier, index) => {
+
+            // first tier has no previous tier to upgrade from
+            if(index == 0) return;
+
+            const prev = black_hole_tiers[index - 1];
+            recipes.push({
+                input1: `industrialforegoing:${prev}_black_hole_${type}`,
+                input2: `industrialforegoing:machine_frame_${tier}`,
+                output: `industrialforegoing:${tier}_black_hole_${type}`,
+                id: `industrialforegoing:${tier}_black_hole_${type}_smithing`
+            });
+        });
+    })
+
     recipes.forEach((recipe) => {
-        event.smithing(recipe.output, recipe.input1, recipe.input2);
+        event.smithing(recipe.output, recipe.input1, recipe.input2).id(recipe.id);
```

Comment:
For the ID's, I'd like to keep the same format as other places.

At the top of the file, there should be an id_prefix: `const id_prefix = 'enigmatica:expert/smithing/';`

and the id in the recipe would then follow the format of id: ${id_prefix} `${id_prefix}{tier}_black_hole_${type}_smithing`

Reply:
Since all the other smithing recipes don't use id prefixes either, mind if i request this to be merged as-is, and then a separate pr to prefix all the smithing recipes properly?

Reply:
and if the prefix has `/smithing/` in it, will it still need to end with `_smithing`?

Reply:
Imo, if anything you'd want it to be like `upgrade` or something.

Reply:
Also here's a question - is there any reason we don't just let them jump up multiple tiers like this? Seems fine to go from Pity -> Supreme...